reasoner/resolution/framework/Request.java

```diff
@@ -127,39 +123,67 @@ public String toString() {
     }
 
     public static class Path {
-        final List<Actor<? extends Resolver<?>>> path;
 
-        public Path(Actor<? extends Resolver<?>> sender) {
-            this(list(sender));
+        private final List<VisitedResolver> path;
+
+        public Path(Actor<? extends Resolver<?>> sender, AnswerState.DownstreamVars answerState) {
+            this(list(new VisitedResolver(sender, answerState)));
         }
 
-        private Path(List<Actor<? extends Resolver<?>>> path) {
-            assert !path.isEmpty() : "Path cannot be empty";
+        public Path(List<VisitedResolver> path) {
             this.path = path;
         }
 
-        public Path append(Actor<? extends Resolver<?>> actor) {
-            List<Actor<? extends Resolver<?>>> appended = new ArrayList<>(path);
-            appended.add(actor);
+        public Path append(Actor<? extends Resolver<?>> actor, AnswerState.DownstreamVars answerState) {
+            List<VisitedResolver> appended = new ArrayList<>(path);
+            appended.add(new VisitedResolver(actor, answerState));
             return new Path(appended);
         }
 
         @Override
         public boolean equals(Object o) {
             if (this == o) return true;
             if (o == null || getClass() != o.getClass()) return false;
-            Path path1 = (Path) o;
-            return Objects.equals(path, path1.path);
+            Path other = (Path) o;
+            return Objects.equals(path, other.path);
         }
 
         @Override
         public int hashCode() {
             return Objects.hash(path);
         }
 
-        public Actor<RootResolver> root() {
-            assert path.get(0).state instanceof RootResolver;
-            return (Actor<RootResolver>) path.get(0);
+        public Actor<? extends Resolver<?>> root() {
+            assert path.get(0).resolver.state instanceof Root || path.get(0).resolver.state instanceof NegationResolver;
+            return path.get(0).resolver;
+        }
+
+        /**
+         * To distinguish between visiting a resolver with two different unifiers,
+         */
+        private static class VisitedResolver {
```

Comment:
this, along with `Path` will be removed once the `AnswerState` has been upgraded to take care of all equality checks we require!